controllers/class.tapestry-controller.php

```diff
@@ -0,0 +1,337 @@
+        }
+
+        $tapestry = get_post_meta($this->postId, 'tapestry', true);
+
+        if (!isset($tapestry->groups)) {
+            $tapestry->groups = [];
+        }
+
+        array_push($tapestry->groups, $group->id);
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+
+        return $group;
+    }
+
+    /**
+     * Get the node controller with associated node meta ID
+     * 
+     * @param   Number  $nodeMetaId node meta ID
+     *
+     * @return  Object  $node       node controller
+     */
+    public function getNode($nodeMetaId)
+    {
+        return new TapestryNodeController($this->postId, $nodeMetaId);
+    }
+
+    /**
+     * Get the group controller
+     * 
+     * @return  Object  $group  group controller
+     */
+    public function getGroups()
+    {
+        return new TapestryGroupController($this->postId);
+    }
+
+    /**
+     * Get the link controller
+     * 
+     * @return  Object  $link   link controller
+     */
+    public function getLinks()
+    {
+        return new TapestryLinkController($this->postId);
+    }
+
+    /**
+     * Get the setting controller
+     * 
+     * @return  Object  $setting   setting controller
+     */
+    public function getSettings()
+    {
+        return new TapestrySettingController($this->postId);
+    }
+
+    private function _addTapestry($tapestry)
+    {
+        $this->postId = TapestryHelpers::updatePost($tapestry, 'tapestry');
+
+        $tapestry->links = [];
+        $tapestry->nodes = [];
+        $tapestry->groups = [];
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+
+        return $tapestry;
+    }
+
+    private function _formNodeData($nodeData, $nodeMetadata)
+    {
+        // Update node data here to match its own version
+        // This enables the same node to have multiple versions
+        $nodeData->id = (int) $nodeMetadata->meta_id;
+        if (isset($nodeMetadata->meta_value->title)) {
+            $nodeData->title = $nodeMetadata->meta_value->title;
+        }
+        if (isset($nodeMetadata->meta_value->coordinates->x)) {
+            $nodeData->fx = $nodeMetadata->meta_value->coordinates->x;
+        }
+        if (isset($nodeMetadata->meta_value->coordinates->y)) {
+            $nodeData->fy = $nodeMetadata->meta_value->coordinates->y;
+        }
+        if (isset($nodeMetadata->meta_value->permissions)) {
+            $nodeData->permissions = $nodeMetadata->meta_value->permissions;
+        }
+        if (isset($nodeMetadata->meta_value->typeData)) {
+            $nodeData->typeData = $nodeMetadata->meta_value->typeData;
+        }
+        if (isset($nodeMetadata->meta_value->imageURL)) {
+            $nodeData->imageURL = $nodeMetadata->meta_value->imageURL;
+        }
+        if (isset($nodeMetadata->meta_value->unlocked)) {
+            $nodeData->unlocked = $nodeMetadata->meta_value->unlocked;
+        }
+        return $nodeData;
+    }
+
+    private function _getTapestry()
```

Comment:
I would change the name of this function to `_loadFromDatabase` and then make sure you are saving the info you are retrieving to the class private properties. You can then call this function from your constructor.